component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,340 @@
+import static org.wso2.carbon.identity.oauth2.grant.token.exchange.utils.TokenExchangeUtils.getIDPAlias;
+import static org.wso2.carbon.identity.oauth2.grant.token.exchange.utils.TokenExchangeUtils.handleCustomClaims;
+import static org.wso2.carbon.identity.oauth2.grant.token.exchange.utils.TokenExchangeUtils
+        .parseTokenExchangeConfiguration;
+import static org.wso2.carbon.identity.oauth2.grant.token.exchange.utils.TokenExchangeUtils.handleException;
+import static org.wso2.carbon.identity.oauth2.grant.token.exchange.utils.TokenExchangeUtils.setAuthorizedUser;
+import static org.wso2.carbon.identity.oauth2.grant.token.exchange.utils.TokenExchangeUtils.validateSignature;
+import static org.wso2.carbon.identity.oauth2.util.OAuth2Util.isJWT;
+
+/**
+ * Class to handle Token Exchange grant type.
+ */
+public class TokenExchangeGrantHandler extends AbstractAuthorizationGrantHandler {
+
+    private static final Log log = LogFactory.getLog(TokenExchangeGrantHandler.class);
+    private int validityPeriodInMin;
+    private final String[] registeredClaimNames = new String[]{"iss", "sub", "aud", "exp", "nbf", "iat", "jti"};
+    private String requestedTokenType = Constants.TokenExchangeConstants.JWT_TOKEN_TYPE;
+
+    /**
+     * Initialize the TokenExchangeGrantHandler.
+     *
+     * @throws IdentityOAuth2Exception Error when initializing
+     */
+    public void init() throws IdentityOAuth2Exception {
+
+        super.init();
+        Map<String, String> configMap = parseTokenExchangeConfiguration();
+        setValidityPeriod(configMap.get(Constants.ConfigElements.IAT_VALIDITY_PERIOD_IN_MIN));
+        if (log.isDebugEnabled()) {
+            log.debug("IAT validity period is set to: " + validityPeriodInMin + "minutes for " +
+                    "Token Exchange grant");
+        }
+    }
+
+    /**
+     * Validate the Token Exchange Grant.
+     * Checks whether the token request satisfies the requirements to exchange the token.
+     *
+     * @param tokReqMsgCtx OAuthTokenReqMessageContext
+     * @return true or false if the grant_type is valid or not.
+     * @throws IdentityOAuth2Exception Error when validating the Token Exchange Grant
+     */
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        String requestedAudience = null;
+        RequestParameter[] params = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getRequestParameters();
+        Map<String, String> requestParams = Arrays.stream(params).collect(Collectors.toMap(RequestParameter::getKey,
+                requestParam -> requestParam.getValue()[0]));
+        String subjectTokenType = requestParams.get(Constants.TokenExchangeConstants.SUBJECT_TOKEN_TYPE);
+
+        if (requestParams.get(Constants.TokenExchangeConstants.REQUESTED_TOKEN_TYPE) != null) {
+            requestedTokenType = requestParams.get(Constants.TokenExchangeConstants.REQUESTED_TOKEN_TYPE);
```

Comment:
Shouldn't this be a local variable?

Reply:
it is used to set the `issued_token_type` in the response - [1]. since it's validated here, we can directly use it to set it in the response.

[1] https://github.com/vithu30/identity-oauth2-grant-token-exchange/blob/main/component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java#L142